doc/src/base/scopedvalues.md

```diff
@@ -0,0 +1,220 @@
+values. You might want to explicitly [unshare mutable state](@ref unshare_mutable_state)
+when entering a new dynamic scope.
+
+```julia
+import Base.Threads: @spawn
+const sval_dict = ScopedValue(Dict())
+
+# Example of using a mutable value wrongly
+@sync begin
+    # `Dict` is not thread-safe the usage below is invalid
+    @spawn (sval_dict[][:a] = 3)
+    @spawn (sval_dict[][:b] = 3)
+end
+
+@sync begin
+    # If we instead pass a unique dictionary to each
+    # task we can access the dictonaries race free.
+    scoped(sval_dict => Dict())
+        @spawn (sval_dict[][:a] = 3)
+    end
+    scoped(sval_dict => Dict())
+        @spawn (sval_dict[][:b] = 3)
+    end
+end
+```
+
+## Example
+
+In the example below we use a scoped value to implement a permission check in
+a web-application. After determining the permissions of the request,
+a new dynamic scope is entered and the scoped value `LEVEL` is set.
+Other parts of the application can query the scoped value and will receive
+the appropriate value. Other alternatives like task-local storage and global variables
+are not well suited for this kind of propagation; our only alternative would have
+been to thread a value through the entire call-chain.
+
+```julia
+const LEVEL = ScopedValue(:GUEST)
+
+function serve(request, response)
+    level = isAdmin(request) ? :ADMIN : :GUEST
+    scoped(LEVEL => level) do
+        Threads.@spawn handle(request, respone)
+    end
+end
+
+function open(connection::Database)
+    level = LEVEL[]
+    if level !== :ADMIN
+        error("Access disallowed")
+    end
+    # ... open connection
+end
+
+function handle(request, response)
+    # ...
+    open(Database(#=...=#))
+    # ...
+end
+```
+
+## Idioms
+### [Unshare mutable state](@id unshare_mutable_state)
+
+```julia
+import Base.Threads: @spawn
+const sval_dict = ScopedValue(Dict())
+
+# If you want to add new values to the dict, instead of replacing
+# it, unshare the values explicitly. In this example we use `merge`
+# to unshare the state of the dictonary in parent scope.
+@sync begin
+    scoped(sval_dict => merge(sval_dict, Dict(:a => 10))) do
+        @spawn @show sval_dict[][:a]
+    end
+    @spawn sval_dict[][:a] = 3 # Not a race since they are unshared.
+end
+```
+
+### [Local caching](@id local_caching)
+
+Since lookup of a scoped variable is linear in scope depth, it can be beneficial
+for a library at an API boundary to cache the state of the scoped value.
+
+```julia
+const DEVICE = ScopedValue(:CPU)
+
+function solve_problem(args...)
+    # Cache current device
+    @scoped DEVICE => DEVICE[] begin
+        # call functions that use `DEVICE[]` repeatedly.
+    end
+```
+
+### Scoped values as globals
+
+In order to access the value of a scoped value, the scoped value itself has to
+be in (lexical) scope. This means most often you likely want to used scoped values
+as constant globals.
+
+```julia
+const sval = ScopedValue(1)
```

Comment:
Very minor but I think we'll have questions about it (namespacing is already confusing for a lot of people):

```suggestion
module Foo
export sval1
const sval1 = ScopedValue(1)
const sval2 = ScopedValue(2)
bar() = @show sval1[] sval2[]
end

using Foo
@scoped sval1 => 10 Foo.sval2 => 20 Foo.bar()
```

Reply:
I am not sure the module helps understandability.
Is there a section of the manual we could reference?

Reply:
I think it's important to show how scoping works from both directions (caller and callee, or dev and user). Most often users will be scoping on package specific globals which may or may not be exported. 

Both caller and callee respect normal lexical scoping, so perhaps referencing https://docs.julialang.org/en/v1/manual/variables-and-scoping/ and saying that the `ScopedValue` must be in lexical scope when accessed *and when set* is sufficient.